dm/syncer/data_validator.go

```diff
@@ -447,119 +610,143 @@ func (v *DataValidator) processRowsEvent(header *replication.EventHeader, ev *re
-		columnMap[col.Name.O] = col
-	}
-	pk := table.PrimaryKey
-	pkIndices := make([]int, len(pk.Columns))
-	for i, col := range pk.Columns {
-		pkIndices[i] = columnMap[col.Name.O].Offset
-	}
-	table.pkIndices = pkIndices
-
 	step := 1
 	if changeType == rowUpdated {
 		step = 2
 	}
 	for i := 0; i < len(ev.Rows); i += step {
 		row := ev.Rows[i]
-		pkValue := make([]string, len(pk.Columns))
-		for _, idx := range pkIndices {
+		pkValue := make([]string, len(table.pkIndices))
+		for _, idx := range table.pkIndices {
 			pkValue[idx] = genColData(row[idx])
 		}
 		key := genRowKey(pkValue)
 
 		if changeType == rowUpdated {
 			afterRowChangeType := changeType
 			afterRow := ev.Rows[i+1]
-			afterPkValue := make([]string, len(pk.Columns))
-			for _, idx := range pkIndices {
+			afterPkValue := make([]string, len(table.pkIndices))
+			for _, idx := range table.pkIndices {
 				afterPkValue[idx] = genColData(afterRow[idx])
 			}
 			afterKey := genRowKey(afterPkValue)
 			if afterKey != key {
 				// TODO: may reuse IsIdentityUpdated/SplitUpdate of RowChange in pkg/sqlmodel
 				v.dispatchRowChange(key, &rowChange{
-					table:      table,
-					key:        key,
-					pkValues:   pkValue,
-					data:       row,
-					tp:         rowDeleted,
-					lastMeetTS: int64(header.Timestamp),
+					table: table,
+					Key:   key,
+					Data:  row,
+					Tp:    rowDeleted,
 				})
 				afterRowChangeType = rowInsert
 			}
 			v.dispatchRowChange(afterKey, &rowChange{
-				table:      table,
-				key:        afterKey,
-				pkValues:   afterPkValue,
-				data:       afterRow,
-				tp:         afterRowChangeType,
-				lastMeetTS: int64(header.Timestamp),
+				table: table,
+				Key:   afterKey,
+				Data:  afterRow,
+				Tp:    afterRowChangeType,
 			})
 		} else {
 			v.dispatchRowChange(key, &rowChange{
-				table:      table,
-				key:        key,
-				pkValues:   pkValue,
-				data:       row,
-				tp:         changeType,
-				lastMeetTS: int64(header.Timestamp),
+				table: table,
+				Key:   key,
+				Data:  row,
+				Tp:    changeType,
 			})
 		}
 	}
 	return nil
 }
 
+func (v *DataValidator) flushCheckpointAndData(loc binlog.Location) error {
+	var wg sync.WaitGroup
+	wg.Add(v.workerCnt)
+	flushJob := &rowChange{
+		Tp: flushCheckpoint,
+		wg: &wg,
+	}
+	for _, worker := range v.workers {
+		worker.rowChangeCh <- flushJob
+	}
+	wg.Wait()
+
+	err := v.persistHelper.persist(loc)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func (v *DataValidator) loadPersistedData(tctx *tcontext.Context) error {
+	var err error
+	v.location, err = v.persistHelper.loadCheckpoint(tctx)
+	if err != nil {
+		return err
+	}
+
+	var rev int64
+	v.loadedPendingChanges, rev, err = v.persistHelper.loadPendingChange(tctx)
+	if err != nil {
+		return err
+	}
+	// fix missing data
+	for _, tblChange := range v.loadedPendingChanges {
+		tblChange.table, err = v.genValidateTableInfo(tblChange.table.Source)
+		if err != nil {
+			return err
```

Comment:
in line 614 we ignored some error. But same error will cause whole routine failed here.

Reply:
in 614, we just met the row and has to decide whether this row should be validated, so if the err shows this row we cannot validate, it's skipped. here this row should be validated(already persisted), so should not have error when getting table info.

Reply:
maybe it will happen when

... DML on t1, DROP TABLE t1, DML on other tables, ...

the validator has pending changes of DML on t1, but syncer has dropped t1, so validator can't bootup. Maybe multi-version TableInfo can help it, we can fix it later

Reply:
add a todo